modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/UncollectExecutionTest.java

```diff
@@ -0,0 +1,160 @@
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.exec.rel;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.ignite.internal.processors.query.calcite.exec.ExecutionContext;
+import org.apache.ignite.internal.processors.query.calcite.type.IgniteTypeFactory;
+import org.apache.ignite.internal.processors.query.calcite.util.TypeUtils;
+import org.apache.ignite.internal.util.typedef.F;
+import org.junit.Test;
+
+/**
+ * Test for UNCOLLECT node execution.
+ */
+public class UncollectExecutionTest extends AbstractExecutionTest {
+    /** */
+    @Test
+    public void testCollectionSizes() {
+        for (boolean withOrdinality : new boolean[] {false, true}) {
+            int rows = 100;
+
+            int[] sizes = {1, IN_BUFFER_SIZE / 2 - 1, IN_BUFFER_SIZE / 2, IN_BUFFER_SIZE / 2 + 1, IN_BUFFER_SIZE,
+                IN_BUFFER_SIZE + 1, IN_BUFFER_SIZE * 4};
+
+            for (int size : sizes) {
+                log.info("Check: size=" + size + ", withOrdinality=" + withOrdinality);
+
+                Function<Integer, Object> func = row -> IntStream.range(0, size).boxed().collect(Collectors.toList());
+
+                RootRewindable<Object[]> root = createNodes(withOrdinality, new TestTable(rows, 1, func));
+
+                long[] colsSum = new long[withOrdinality ? 2 : 1];
+                int cnt = 0;
+
+                while (root.hasNext()) {
+                    cnt++;
+                    Object[] row = root.next();
+
+                    for (int i = 0; i < row.length; i++)
+                        colsSum[i] += (int)row[i];
+                }
+
+                assertEquals(size * rows, cnt);
+
+                // Expected value of unnested column = rows * sum(0 .. size - 1)
+                assertEquals("Unexpected sum: size=" + size, (long)size * (size - 1) / 2 * rows, colsSum[0]);
+
+                if (withOrdinality) // Expected value of ordinality column = rows * sum(1 .. size)
+                    assertEquals("Unexpected ordinality: size=" + size, (long)size * (size + 1) / 2 * rows, colsSum[1]);
+
+                root.closeRewindableRoot();
+            }
+        }
+    }
+
+    /** */
+    @Test
+    public void testExactResult() {
+        checkExactResult(
+            false,
+            F.asList(
+                row(F.asList(1, 2)),
+                row(F.asList(1))
+            ),
+            F.asList(
+                row(1), row(2), row(1)
+            )
+        );
+
+        checkExactResult(
+            true,
+            F.asList(
+                row(F.asList(1, 2)),
+                row(F.asList(3, 4, 5)),
+                row(F.asList(6, 7))
+            ),
+            F.asList(
```

Comment:
Wierd ordinalinty. Why not row number? It sutes for result like:

      1 |      3 |      6 |          1
      2 |      4 |      7 |          2
         |      5 |         |          3